# Hybrid Retrieval and Reranking

The final inference pipeline. This notebook implements a Two-Stage Architecture:

Retrieval (Recall Layer): A Hybrid approach combining Dense Vectors (Matryoshka 64d) and Neural Sparse (SPLADE + BM25) to maximize candidate coverage.

Reranking (Precision Layer): A Cross-Encoder (mxbai-rerank) re-scores the top candidates to maximize nDCG.

In [1]:
import sys, os, yaml, time, torch, pickle, dagshub, mlflow
import pandas as pd
from pathlib import Path
import numpy as np
from sentence_transformers import CrossEncoder

# Path Setup
project_root = Path("..").resolve()
if str(project_root) not in sys.path:
    sys.path.append(str(project_root))
os.chdir(project_root)
    
from src.esci.sparse_retrievers import SPLADEFast, BM25Fast
from src.esci.system_b import build_candidates, rerank_candidates
from src.esci.metrics import compute_recall_metrics, compute_ndcg_metrics
from src.esci.artifacts import load_artifacts
from src.esci.mlflow import log_candidates_run, log_rerank_run

# Load Config
with open(project_root / "configs" / "esci.yaml") as f:
    cfg = yaml.safe_load(f)

## Setting up MLflow connection

In [2]:
dagshub.init(repo_owner='aminlasri', repo_name='Amazon-Search-Engine-Project', mlflow=True)

Accessing as aminlasri

Initialized MLflow to track repo "aminlasri/Amazon-Search-Engine-Project"

Repository aminlasri/Amazon-Search-Engine-Project initialized!

In [3]:
print("Tracking URI:", mlflow.get_tracking_uri())

Tracking URI: https://dagshub.com/aminlasri/Amazon-Search-Engine-Project.mlflow


Here, we are setting the experiment name based on the variable hyperparameters.

In [4]:
mlflow.set_experiment("XP_1_rrf_k_60_weights_splade_0.5_bm25_0.3")

<Experiment: artifact_location='mlflow-artifacts:/bf947ca8c30b492a89665d3b14dd320c', creation_time=1774021376594, experiment_id='0', last_update_time=1774021376594, lifecycle_stage='active', name='XP_1_rrf_k_60_weights_splade_0.5_bm25_0.3', tags={}>

## STAGE 1: Hybrid Candidate Generation

In [5]:
# Strategy: Dense (64d) + SPLADE + BM25 (Hybrid)
print("Setting up Retrieval Phase...")

# Load Artifacts
prod_emb, qry_emb, prod_df, qry_df, pair_df = load_artifacts("us", "test", cfg)

# Build Ground Truth (q_rels)
print("Building Ground Truth Map (q_rels)...")
q_rels = pair_df.groupby("query_id")["grade"].apply(list).to_dict()

# Load SPLADE
print("Loading SPLADE Matrix...")
splade_path = "artifacts/systemA/splade_data.pt"
splade_data = torch.load(splade_path, map_location="cpu")

splade_loaded = SPLADEFast(
    cfg["sparse"]["splade_model"], 
    batch_size=cfg["sparse"]["batch_size"]
)
splade_loaded.doc_matrix = splade_data["doc_matrix"]
splade_loaded.pid_list   = splade_data["pid_list"]

# Align device
splade_loaded.doc_matrix = splade_loaded.doc_matrix.to(splade_loaded.device)
splade = splade_loaded

print("Loading BM25 Index...")
with open("artifacts/systemA/bm25_retriever.pkl", "rb") as f:
    bm25 = pickle.load(f)

# Configure Retrieval
# Dense + SPLADE + BM25
cfg["retrieval"]["sources"] = ["dense", "splade", "bm25"] 
dim_to_use = 64
sources = cfg["retrieval"]["sources"]

print(f"Retrieving Candidates (Matryoshka Dim {dim_to_use} + BM25 + SPLADE)...")

retrieval_df, retrieval_qps = build_candidates(
    cfg, 
    split="test", 
    override_dim=dim_to_use, 
    prebuilt_splade=splade, 
    prebuilt_bm25=bm25  
)

# Measure Retrieval Baseline
recall_results = compute_recall_metrics(retrieval_df, q_rels)
ndcg_results = compute_ndcg_metrics(retrieval_df, q_rels)

# Merge metrics
base_metrics = {**recall_results, **ndcg_results}

# save results to mlflow
log_candidates_run(cfg, dim_to_use, sources, base_metrics, retrieval_qps)

print(f"\n Retrieval Dense + SPLADE + BM25 (Dim {dim_to_use}):")
print(f"   nDCG@10:    {base_metrics['nDCG@10']:.4f}")
print(f"   nDCG@20:    {base_metrics['nDCG@20']:.4f}")
print(f"   nDCG@50:    {base_metrics['nDCG@50']:.4f}")
print(f"   Recall@100: {base_metrics['Recall@100']:.4f}")
print(f"   Recall@200: {base_metrics['Recall@200']:.4f}")
print(f"   QPS:        {retrieval_qps:.2f}")

Setting up Retrieval Phase...
Building Ground Truth Map (q_rels)...
Loading SPLADE Matrix...


C:\Users\aminl\anaconda3\envs\pytorch-gpu\Lib\site-packages\transformers\tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


Loading BM25 Index...
Retrieving Candidates (Matryoshka Dim 64 + BM25 + SPLADE)...
 Using FAISS (Dim=64)
    -> Measuring QPS for 8908 queries...
🏃 View run candidates__dim64__dense_splade_bm25 at: https://dagshub.com/aminlasri/Amazon-Search-Engine-Project.mlflow/#/experiments/0/runs/f4e778128b6f4bf1a8c5f5773dc0311d
🧪 View experiment at: https://dagshub.com/aminlasri/Amazon-Search-Engine-Project.mlflow/#/experiments/0

 Retrieval Dense + SPLADE + BM25 (Dim 64):
   nDCG@10:    0.4470
   nDCG@20:    0.4846
   nDCG@50:    0.5493
   Recall@100: 0.7406
   Recall@200: 0.8148
   QPS:        69.63


### Summary of Results from the 7 Experiments

In [12]:
mlflow_results = pd.read_csv("results/mlflow_results.csv", index_col=0)
mlflow_results

,1,2,3,4,5,6,7
Experiment,,,,,,,
rrf_k,60.000,60.000,60.000,60.000,80.000,40.000,20.000
w_bm25,0.300,0.300,0.200,0.300,0.300,0.300,0.300
w_splade,0.500,0.700,0.500,0.300,0.500,0.500,0.500
QPS,69.090,72.490,73.480,71.100,72.090,71.420,72.520
Recall_100,0.741,0.744,0.738,0.729,0.738,0.742,0.742
Recall_200,0.815,0.813,0.814,0.811,0.815,0.814,0.813
Recall_50,0.636,0.646,0.629,0.614,0.632,0.640,0.644
nDCG_10,0.447,0.460,0.443,0.435,0.447,0.446,0.446
nDCG_20,0.485,0.500,0.479,0.470,0.484,0.486,0.489


#### Final Choice: rrf:60 , Dense:1.0 , BM25: 0.3 , SPLADE:0.5 (Experiment 1)
#### WHY?

The selection of these hyperparameters is justified by the need of a higher Recall@200. The current reranker focuses on the top 50 products, but maintaining a high Recall@200 ensures the system is optimized for future hardware upgrades and expanded reranking capabilities.
- To see all the 7 Experiments visit: https://dagshub.com/aminlasri/Amazon-Search-Engine-Project.mlflow/#/experiments

## A function to help find the best batch size for the reranker

In [6]:
def autotune_batch_size(ce, pairs, batch_sizes=(4,8,16,32,64,128), warmup=1):
    device = ce.device
    results = []
    
    # small slice for tuning
    sample = pairs[:min(len(pairs), 2000)]
    
    for bs in batch_sizes:
        try:
            torch.cuda.empty_cache()
            # warmup
            for _ in range(warmup):
                _ = ce.predict(sample[:bs], batch_size=bs, show_progress_bar=False, num_workers=0)

            t0 = time.time()
            _ = ce.predict(sample, batch_size=bs, show_progress_bar=False, num_workers=0)
            dt = time.time() - t0
            
            qps = len(sample) / dt
            mem = torch.cuda.max_memory_allocated() / (1024**3) if torch.cuda.is_available() else 0.0
            results.append((bs, qps, mem))
            print(f"bs={bs:4d} | QPS={qps:8.1f} | peak_mem={mem:.2f} GB")
        except RuntimeError as e:
            if "out of memory" in str(e).lower():
                print(f"bs={bs:4d} | OOM")
            else:
                raise e

    if not results:
        return None, []
    best = max(results, key=lambda x: x[1])
    return best[0], results


In [7]:
ce_model = cfg["cross_encoder_model"]

ce = CrossEncoder(
    ce_model,
    device="cuda",
    max_length= cfg["reranker"]["max_seq_length"]
)

pairs = list(zip(
    retrieval_df["query"].head(2000),
    retrieval_df["product_text_dense"].head(2000)
))

best_bs, res = autotune_batch_size(
    ce,
    pairs,
    batch_sizes=(4,8,16,32,64,128)
)

print("Best batch_size:", best_bs)

The CrossEncoder.predict `num_workers` argument is deprecated and has no effect. It will be removed in a future version.
The CrossEncoder.predict `num_workers` argument is deprecated and has no effect. It will be removed in a future version.
The CrossEncoder.predict `num_workers` argument is deprecated and has no effect. It will be removed in a future version.
The CrossEncoder.predict `num_workers` argument is deprecated and has no effect. It will be removed in a future version.


bs=   4 | QPS=   204.5 | peak_mem=1.47 GB


The CrossEncoder.predict `num_workers` argument is deprecated and has no effect. It will be removed in a future version.
The CrossEncoder.predict `num_workers` argument is deprecated and has no effect. It will be removed in a future version.


bs=   8 | QPS=   270.3 | peak_mem=1.52 GB


The CrossEncoder.predict `num_workers` argument is deprecated and has no effect. It will be removed in a future version.
The CrossEncoder.predict `num_workers` argument is deprecated and has no effect. It will be removed in a future version.


bs=  16 | QPS=   281.5 | peak_mem=1.63 GB


The CrossEncoder.predict `num_workers` argument is deprecated and has no effect. It will be removed in a future version.


bs=  32 | QPS=   263.7 | peak_mem=1.86 GB


The CrossEncoder.predict `num_workers` argument is deprecated and has no effect. It will be removed in a future version.
The CrossEncoder.predict `num_workers` argument is deprecated and has no effect. It will be removed in a future version.


bs=  64 | QPS=   247.2 | peak_mem=2.32 GB


The CrossEncoder.predict `num_workers` argument is deprecated and has no effect. It will be removed in a future version.


bs= 128 | QPS=   248.7 | peak_mem=3.25 GB
Best batch_size: 16


## STAGE 2: Cross-Encoder Reranking

In [8]:
ce_model = cfg["cross_encoder_model"]

print(f"\n Reranking Top 50 Candidates with {ce_model}...")

# Rerank the top 50 candidates
reranked_df, rerank_qps = rerank_candidates(
    retrieval_df, 
    model_name=ce_model, 
    top_k_to_rerank=50,
    batch_size=best_bs,
    cfg=cfg)

# Measure Reranking Performance
final_metrics = compute_ndcg_metrics(reranked_df, q_rels)

# Save results to mlflow
log_rerank_run(cfg, dim_to_use, sources, final_metrics, rerank_qps)

print(f"\n Reranking Performance:")
print(f"   nDCG@10:    {final_metrics['nDCG@10']:.4f}")
print(f"   nDCG@20:    {final_metrics['nDCG@20']:.4f}")
print(f"   nDCG@50:    {final_metrics['nDCG@50']:.4f}")
print(f"   QPS:        {rerank_qps:.2f}")


 Reranking Top 50 Candidates with mixedbread-ai/mxbai-rerank-base-v1...
 Loading Cross-Encoder: mixedbread-ai/mxbai-rerank-base-v1 on CUDA... and Max length: 128
    -> Reranking 445400 pairs...
    -> Rerank Speed: 5.72 QPS
🏃 View run rerank__dim64__dense_splade_bm25 at: https://dagshub.com/aminlasri/Amazon-Search-Engine-Project.mlflow/#/experiments/0/runs/535d072a697c44ff9461d0f486a3a291
🧪 View experiment at: https://dagshub.com/aminlasri/Amazon-Search-Engine-Project.mlflow/#/experiments/0

 Reranking Performance:
   nDCG@10:    0.5064
   nDCG@20:    0.5378
   nDCG@50:    0.5848
   QPS:        5.72


## STEP 3: The Comparison

In [9]:
comparison = pd.DataFrame([
    {
        "Stage": "1. Retrieval (Hybrid 64d)",
        "nDCG@10": base_metrics["nDCG@10"],
        "nDCG@20": base_metrics["nDCG@20"],
        "nDCG@50": base_metrics["nDCG@50"],
        "Recall@100": base_metrics["Recall@100"],
        "Recall@200": base_metrics["Recall@200"],
        "QPS": retrieval_qps
    },
    {
        "Stage": "2. Reranking (Cross-Encoder)",
        "nDCG@10": final_metrics["nDCG@10"],
        "nDCG@20": final_metrics["nDCG@20"],
        "nDCG@50": final_metrics["nDCG@50"],
        "QPS": rerank_qps
    }
])

print("\nFinal Pipeline Results:")
display(comparison)

# Calculate Accuracy Lift (nDCG is the main metric for ranking quality)
lift10 = (final_metrics['nDCG@10'] - base_metrics['nDCG@10']) / base_metrics['nDCG@10'] * 100
print(f"\n Accuracy @10 Lift from Reranking: +{lift10:.1f}%")

lift20 = (final_metrics['nDCG@20'] - base_metrics['nDCG@20']) / base_metrics['nDCG@20'] * 100
print(f"\n Accuracy @20 Lift from Reranking: +{lift20:.1f}%")

lift50 = (final_metrics['nDCG@50'] - base_metrics['nDCG@50']) / base_metrics['nDCG@50'] * 100
print(f"\n Accuracy @50 Lift from Reranking: +{lift50:.1f}%")


Final Pipeline Results:


,Stage,nDCG@10,nDCG@20,nDCG@50,Recall@100,Recall@200,QPS
0,1. Retrieval (Hybrid 64d),0.446964,0.484621,0.549283,0.740644,0.814804,69.632147
1,2. Reranking (Cross-Encoder),0.506373,0.537757,0.584781,NaN,NaN,5.715048



 Accuracy @10 Lift from Reranking: +13.3%

 Accuracy @20 Lift from Reranking: +11.0%

 Accuracy @50 Lift from Reranking: +6.5%


By using JUST BASIC(base) models: "BAAI/bge-base-en-v1.5" for RETRIEVAL and "mixedbread-ai/mxbai-rerank-base-v1" for RERANKING, we managed to get acceptable results: Recall@200 = 81.48% and nDCG@20 = 53.78%.

With access to enterprise compute (e.g., NVIDIA A100s or H100s), these metrics can be aggressively pushed higher through three direct upgrades:
- Parameter Scaling with Larger models.
- Larger Batch Sizes which will increase the discriminative power of MNRL.
- Expanded Context Windows by increasing the max_seq_length to 512 or 1024.